<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>My Project: predict_promoters.py File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">My Project
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<div class="title">predict_promoters.py File Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>This file implements function that builds and trains neural network for TSS prediction using specified hyper-parameters, it converges after ~20 epochs reaching ~95% accuracy.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:adfc6c6e4919d2b6fe2a086f4f4fa49a4"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="predict__promoters_8py.html#adfc6c6e4919d2b6fe2a086f4f4fa49a4">predict_promoters.buid_model_and_train</a> (use_inputs, data, epochs=10, l1=0.0, l2=0.01, dense=[256], optimizer=&quot;Adam&quot;, learning_rate=0.001, model=None, distance=5, activation=&quot;relu&quot;)</td></tr>
<tr class="separator:adfc6c6e4919d2b6fe2a086f4f4fa49a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6ed239a522fc59550787770204603f49"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="predict__promoters_8py.html#a6ed239a522fc59550787770204603f49">predict_promoters.main</a> ()</td></tr>
<tr class="separator:a6ed239a522fc59550787770204603f49"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:a57c0d0efd540b154071f309d0e6c02fe"><td class="memItemLeft" align="right" valign="top"><a id="a57c0d0efd540b154071f309d0e6c02fe"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>predict_promoters.test_size</b> = 8000</td></tr>
<tr class="separator:a57c0d0efd540b154071f309d0e6c02fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6092fd065f7f92afdbf8d0cb2c34a600"><td class="memItemLeft" align="right" valign="top"><a id="a6092fd065f7f92afdbf8d0cb2c34a600"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>predict_promoters.TSS_pos</b> = 500</td></tr>
<tr class="separator:a6092fd065f7f92afdbf8d0cb2c34a600"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>This file implements function that builds and trains neural network for TSS prediction using specified hyper-parameters, it converges after ~20 epochs reaching ~95% accuracy. </p>
</div><h2 class="groupheader">Function Documentation</h2>
<a id="file_adfc6c6e4919d2b6fe2a086f4f4fa49a4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_adfc6c6e4919d2b6fe2a086f4f4fa49a4">&#9670;&nbsp;</a></span>buid_model_and_train()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def predict_promoters.buid_model_and_train </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>use_inputs</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>data</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>epochs</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>l1</em> = <code>0.0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>l2</em> = <code>0.01</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>dense</em> = <code>[256]</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>optimizer</em> = <code>&quot;Adam&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>learning_rate</em> = <code>0.001</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>model</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>distance</em> = <code>5</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>activation</em> = <code>&quot;relu&quot;</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">This function builds tensorflow.keras model and trains it on given input data.

Params:
 - use_inputs: list of input types to be used for TSS detection. Chsoose from:
   "seq":       use DNA sequence (including one 1D convolution)
   "seq-2conv": also use second 1D convolution on the DNA sequence
   "CG":        use CG-skew data via 1D convolution layer
   "CG-avg":    use CG-skew data via simlpe average pooling (no convolution)
   "CA":        use presence of CA di-nucleotide (directm only in &lt;-20;20&gt; interval)
   "COV":       use Rna-seq coverage data
   "METH":      use methylation data (average pooling)
   "SNPs":      use SNP data (average pooling)
 
 - data: dictionary containing input data in form of pandas.Dataframe(s)
 - hidden: list of integers. Length of list = number of hidden dense layers. Each number
           in the list defines size of a hidden layer
 - epochs: number of epochs we want to run (in model.fit())
 - l1: l1 regularization coefficient
 - l2: l2 regularization coefficient
 - optimizer: optimizer type to use. Choose one of: ["Adam", "SGD", "RMSprop", "Adagrad", "Adadelta", "Adamax", "Nadam"]
              description of optimizers here: https://keras.io/optimizers/
 - activation: activation function for dense layers.
</pre> 
</div>
</div>
<a id="file_a6ed239a522fc59550787770204603f49"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_a6ed239a522fc59550787770204603f49">&#9670;&nbsp;</a></span>main()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def predict_promoters.main </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">When this script is ran as main program, it constructs neural network model using
hyper-parameters defined here, and train the network on the input data.
It expects one argument: filename of table, where all input datasets are defined.
Read more in documentation of function read_input_files
</pre> 
</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
